feature #662 #1183

joelpop opened the pull request:
## Description

Fixes #662 & #1183 (partially)

Adds some selectors to `ElementQuery` (integration testing) to parallel those provided by `ComponentQuery` (unit testing). The selectors added in this change are limited to the ones that are "attribute" based. 

- `ElementQuery<T> withAttribute(String attribute)`
- `ElementQuery<T> withAttribute(String attribute, String value)`
- `ElementQuery<T> withAttributeContaining(String attribute, String value)`
- `ElementQuery<T> withoutAttribute(String attribute)`
- `ElementQuery<T> withoutAttribute(String attribute, String value)`
- `ElementQuery<T> withoutAttributeContaining(String attribute, String value)`
- `ElementQuery<T> withId(String id)`
- `ElementQuery<T> withClassName(String... className)`
- `ElementQuery<T> withoutClassName(String... className)`
- `ElementQuery<T> withTheme(String theme)`
- `ElementQuery<T> withoutTheme(String theme)`
- `T single()`

In addition, three existing selectors that did not match the naming convention of the `ComponentQuery` (unit testing) selectors were replaced by ones with matching names and the old methods were deprecated.

- `ElementQuery<T> hasAttribute(String name)` → `ElementQuery<T> withAttribute(String attribute)`
- `ElementQuery<T> attribute(String name, String value)` → `ElementQuery<T> withAttribute(String attribute, String value)`
- `ElementQuery<T> attributeContains(String name, String token)` → `ElementQuery<T> withAttributeContaining(String attribute, String value)`

Added both unit and integration tests for these selectors and others that were missing.

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

Additional missing selectors are currently being considered, such as:

- `ElementQuery<T> withCondition(Predicate<T> condition)`
- `ElementQuery<T> withCaption(String caption)`
- `ElementQuery<T> withCaptionContaining(String text)`
- `ElementQuery<T> withText(String text)`
- `ElementQuery<T> withTextContaining(String text)`
- `<V> ElementQuery<T> withValue(V value)`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/testbench?pullRequest=1774) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/testbench?pullRequest=1774) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/testbench?pullRequest=1774) it.</sub>

joelpop commented:
I want to point out one thing I changed in the implementation of `ElementQuery::id` that deserves some discussion.

Up until now there was no `ElementQuery::single` method or `ElementQuery::withId` method. Because of that, `ElementQuery::id` used `ElementQuery::first`. However, in one of the test cases (`searchShadowDomBeforeLight`) there were intentionally two ids to be found, so `ElementQuery::first` returned just the first one, and all was happy. However, I have implemented `ElementQuery::single` and `ElementQuery::withId` and reimplemented `ElementQuery::id` to be
```
public T id(String id) {
    return withId(id).single();
}
```
to mirror how `ComponentQuery::id` is implemented for unit tests. That change subsequently caused the `searchShadowDomBeforeLight` integration test to fail. I have changed the test from selecting on `.id(id)` to selecting on `.withId(id).first()`.

What I want to make sure of is that this is not breaking anything in valid Vaadin light/shadow DOM HTML. It would be simple enough to switch `ComponentQuery::id`'s implementation back to `.withId().first()`, but that seems wrong, because ids _should_ be unique on a page.

However, this could potentially break some customers' tests, but that might be a valuable discovery for them—that they have duplicate ids.

I have added a note to `ComponentQuery::id`'s Javadoc explaining the new behavior and how to revert to the previous behavior, if desired. My guess is that this change will have extremely low abrasion, and the silver lining is if the change does break something, it will break it for the good.

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -138,69 +144,242 @@ public ElementQuery(Class<T> elementClass, String tagName) {
      */
-    public T id(String id) {
-        return attribute("id", id).first();
+    @Deprecated(since = "9.2")
```
I suppose this change will go to the next minor
```suggestion
    @Deprecated(since = "9.3")
```

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -138,69 +144,242 @@ public ElementQuery(Class<T> elementClass, String tagName) {
-    public ElementQuery<T> hasAttribute(String name) {
-        attributes.add(new AttributeMatch(name));
-        return this;
+    @Deprecated(since = "9.2")
```
```suggestion
    @Deprecated(since = "9.3")
```

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -138,69 +144,242 @@ public ElementQuery(Class<T> elementClass, String tagName) {
+     *
+     * @deprecated use {@link #withAttributeContaining}
+     */
+    @Deprecated(since = "9.2")
```
```suggestion
    @Deprecated(since = "9.3")
```

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -138,69 +144,242 @@ public ElementQuery(Class<T> elementClass, String tagName) {
+     * Selects on elements with the given attribute having the given value.
+     * <p>
+     * For matching a token within the attribute, see
+     * {@link #withAttributeContaining(String, String)}.
```
It could be nice to also add a `@see` tag for the related method

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -138,69 +144,242 @@ public ElementQuery(Class<T> elementClass, String tagName) {
      * <p>
      * For matching the full attribute value, see
-     * {@link #attribute(String, String)}.
+     * {@link #withAttribute(String, String)}.
```
It could be nice to also add a `@see` tag for the related method

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -138,69 +144,242 @@ public ElementQuery(Class<T> elementClass, String tagName) {
+     * <code>class='someclass myclass'</code>.
+     * <p>
+     * For matching the full attribute value, see
+     * {@link #withoutAttribute(String, String)}.
```
It could be nice to also add a `@see` tag for the related method

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -233,6 +412,42 @@ protected SearchContext getContext() {
+    public T single() {
+        List<T> all = all();
+        if (all.size() != 1) {
+            throw new NoSuchElementException(getNoSuchElementMessage(null));
```
I would use different messages when no elements are found and when the size is greater than one.
The message provided by `getNoSuchElementMessage(null)` is ok when no elements are found, but if we get more than one, we should not tell the developer that `No element with tag ... found`.
In that case, I would expect something like `Multiple elements with tag ... found`

joelpop replied:
Good catch!

mcollovati commented on vaadin-testbench-shared/src/main/java/com/vaadin/testbench/ElementQuery.java:
```diff
@@ -427,35 +640,29 @@ private static String getTagName(Class<?> elementClass) {
             Class<? extends TestBenchElement> elementClass) {
         Attribute[] attrs = elementClass.getAnnotationsByType(Attribute.class);
-        if (attrs == null) {
+        if (attrs.length == 0) {
```
nice catch!

mcollovati reviewed (changes requested):
Good job and great that you updated the README!
However, once all the comments will be addressed, I propose to squash the change into two separated commits, one for the README update and the other for the code changes.

joelpop commented on vaadin-testbench-core-junit5/src/main/java/com/vaadin/testbench/AbstractBrowserTestBase.java:
```diff
@@ -336,7 +336,7 @@ protected boolean hasCssClass(WebElement element, String className) {
      *            the actual element
      */
-    protected static void assertEquals(WebElement expectedElement,
+    protected static void assertEqual(WebElement expectedElement,
```
Yes, the reason this method was renamed was due to its _collision_ with the JUnit assertion. Its name makes it impossible to statically import the JUnit assertion one. Preserving and deprecating would not resolve that issue. As this method is protected, is a temporary workaround, and may no longer be needed (per its Javadoc comment, below), it did not seem like an unreasonable change.
```
     * Can be removed if https://dev.vaadin.com/ticket/18484 is fixed.
```

mcollovati replied:
Not completely sure about renaming. For example, it will break `ShortcutsIT` in Flow repo.
Of course, it will not be a big effort to fix the test.
But if we rename it, this should be mentioned in the release notes.
@mshabarov what do you think?

mshabarov replied:
We're trying to avoid breaking changes as much as possible in minor releases, so I don't appreciate the immediate renaming.

The best way to resolve these collision in my opinion is to deprecate `AbstractBrowserTestBase::assertEquals` method and then at some point remove remove it in the possible new major. However, before we do this, we have to apply unwrapping from `AbstractBrowserTestBase::assertEquals` to the `TestBenchElement::equals` (recommended APi for comparing elements), as otherwise we can loose this functionality.
I'm talking about [this unwrapping](https://github.com/vaadin/testbench/blob/48b384c3be62e7c27bdaa7afb447172fbf051503/vaadin-testbench-core-junit5/src/main/java/com/vaadin/testbench/AbstractBrowserTestBase.java#L343-L350).

Anyway, I think this change would be good to make, but it's not strictly related to the ElementQuery improvements, so I propose to make another PR or ticket for it.

joelpop replied:
> However, before we do this, we have to apply unwrapping from AbstractBrowserTestBase::assertEquals to the TestBenchElement::equals (recommended APi for comparing elements), as otherwise we can loose this functionality.
> I'm talking about [this unwrapping](https://github.com/vaadin/testbench/blob/48b384c3be62e7c27bdaa7afb447172fbf051503/vaadin-testbench-core-junit5/src/main/java/com/vaadin/testbench/AbstractBrowserTestBase.java#L343-L350).
> 

@mshabarov Yes, that unwrapping issue (https://dev.vaadin.com/ticket/18484) has long been resolved, so the `AbstractBrowserTestBase::assertEquals` method can be safely deprecated now and removed in the next major.

<img width="604" alt="image" src="https://github.com/vaadin/testbench/assets/62086947/5a31590c-0543-4c20-9f4d-98d537c4d279">

joelpop replied:
So I will (sadly) revert this part of my change.

joelpop commented on vaadin-testbench-integration-tests-junit5/src/test/java/com/vaadin/tests/ElementQueryIT.java:
```diff
@@ -27,127 +28,276 @@ protected Class<? extends Component> getTestView() {
         openTestURL();
         List<TemplateViewElement> elements = $(TemplateViewElement.class).all();
-        Assertions.assertTrue(elements.get(0) instanceof TemplateViewElement);
+        assertNotNull(elements.get(0));
```
The code quality check says:
<img width="687" alt="image" src="https://github.com/vaadin/testbench/assets/62086947/07e03936-8211-4080-a3a7-9a42ca215fc3">

and its assessment is correct. I accepted the suggested action.

mcollovati replied:
Ok, let's change it.

mcollovati commented on vaadin-testbench-integration-tests/src/test/java/com/vaadin/tests/ElementQueryIT.java:
```diff
@@ -30,7 +31,7 @@ protected Class<? extends Component> getTestView() {
         openTestURL();
         List<TemplateViewElement> elements = $(TemplateViewElement.class).all();
-        Assert.assertTrue(elements.get(0) instanceof TemplateViewElement);
+        assertNotNull(elements.get(0));
```
As in the other comment, let's use assertNotNull

joelpop commented on vaadin-testbench-core-junit5/src/test/java/com/vaadin/testbench/ElementQueryTest.java:
```diff
@@ -8,30 +8,29 @@
-import org.openqa.selenium.SearchContext;
-import org.openqa.selenium.WebDriver;
-import org.openqa.selenium.WebElement;
+import org.openqa.selenium.*;
```
I guess that is an IDE setting used when it auto-organizes imports.

joelpop commented on vaadin-testbench-core-junit5/src/test/java/com/vaadin/testbench/ElementQueryTest.java:
```diff
@@ -8,30 +8,29 @@
+import java.util.stream.Collectors;
 
-public class ElementQueryTest {
+import static org.junit.jupiter.api.Assertions.*;
```
I guess that is an IDE setting used when it auto-organizes imports.

joelpop commented on vaadin-testbench-core-junit5/src/test/java/com/vaadin/testbench/ElementQueryTest.java:
```diff
@@ -352,52 +517,49 @@ public static class MyExtendedAndOverriddenFancyViewElement
-            set.add(t);
-        }
-        return set;
+        return Arrays.stream(ts).collect(Collectors.toSet());
```
Yes.

joelpop commented on vaadin-testbench-core/src/test/java/com/vaadin/testbench/ElementQueryTest.java:
```diff
@@ -366,23 +546,20 @@ public void attributesInherited() {
-            set.add(t);
-        }
-        return set;
+        return Arrays.stream(ts).collect(Collectors.toSet());
```
Yes

mcollovati commented on vaadin-testbench-integration-tests-junit5/vite.generated.ts:
This generated file should probably be removed from VCS, but we can do it in another PR

joelpop replied:
Yes, I agree. I think the .gitignore needs some updating at the same time.

mshabarov commented:
> I want to point out one thing I changed in the implementation of ElementQuery::id that deserves some discussion.

First of all, I agree that fail fast if IDs are duplicated is a expected behaviour for `ElementQuery::id` and also it's better to keep it in sync with `ComponentQuery::id`, so I think we can change that.

I recommend to update `ElementQuery::id` javadoc to highlight this and tell how to get an element with duplicated ID if one wants this for any reason. Also, it probably makes sense to recommend the same in the error message that is thrown by `ElementQuery::id`, if it finds more than one element with the same ID.

joelpop commented:
> > I want to point out one thing I changed in the implementation of ElementQuery::id that deserves some discussion.
> 
> First of all, I agree that fail fast if IDs are duplicated is an expected behaviour for `ElementQuery::id` and also it's better to keep it in sync with `ComponentQuery::id`, so I think we can change that.
> 
> I recommend to update `ElementQuery::id` javadoc to highlight this and tell how to get an element with duplicated ID if one wants this for any reason. Also, it probably makes sense to recommend the same in the error message that is thrown by `ElementQuery::id`, if it finds more than one element with the same ID.

Yes, this commenting has already been done.

mcollovati reviewed (approved):
LGTM.
I would just squash into two commits, one for README and the other for Java changes before merging into main

mcollovati commented:
@mshabarov it looks like "Rebase and merge" is not enabled on this repo.
I'll create a new PR with the README updates commit only, and once that is merged we can rebase this one on top of main.
